### PR TITLE
Recursively compute variables

### LINKED ISSUE #3: Var.vars should accept Vars that are themselves functions of multiple Vars
Currently, if Var is passed a value for `func` when instantiated, it then populates the `vars` attribute with the Var objects that are passed into the given function.  But this is only supported for one level: the Var objects passed to the function must be stand-alone, i.e. don't require a function of their own.

The result is the unnecessary proliferation of functions.  For example, my flux divergence calculations currently have MSE, DSE, c_pT, gz, and general versions.  Each one takes as arguments all the arguments required for the pre-calculation (e.g. temp, hght, and sphum for MSE), computes it by a call to that function, then calls the general function with that output as the argument.

Possible solution: when a Calc object encounters a Var with nested functions such as this, it should, from bottom up, create Calc objects that returns the full timeseries of the specified function.  At the next level up, i.e. which is taking in this computed data as one of its Var objects, the code that loads data from netCDF will have to be bypassed, since you already have the timeseries in hand that you want.

### PR DESCRIPTION
**Pull request by spencerkclark**
Closes #3 

This allows one to compose variable objects whose functions take other computed variables as arguments.  For example:
```python
from aospy import Var

a = Var(name='a')
b = Var(name='b', variables=(a, ), func=lambda x: 2. * x)
c = Var(name='c', variables=(a, b), func=lambda x, y: x + y)
```

### CODE REVIEW

**Comment by spencerkclark** [comment]
Note of course this doesn't deal with saving intermediates to disk (or checking for saved intermediates), which is also discussed in #3, but I think what's here is already a reasonable improvement.

**Comment by spencerahill** [comment]
Wow, this is surprisingly elegant!  The logic and tests look good to me.

One big thing: have we tested this out within a `Calc` yet?  I expect that, given these unit tests, everything will be fine, but I still think it would be good to have an explicit integration test for this.

Besides that, a couple documentation requests:

- Can you give `recursively_compute_variable` a full docstring?
- More substantively, can we highlight this functionality in the docs somewhere?  In scanning through our existing pages, I'm not sure where this fits best.  Perhaps [here](https://aospy.readthedocs.io/en/latest/examples.html#physical-variables) within the Examples page.

File: aospy/utils/io.py [hunk 1]
```diff
@@ -154,9 +154,14 @@ def data_name_gfdl(name, domain, data_type, intvl_type, data_yr,
...
+
+    archive_files = []
+    for f in files_list:
+        if f.startswith('/archive'):
+            archive_files.append(f)
     try:
-        if isinstance(files_list, str):
-            files_list = [files_list]
-        subprocess.call(['dmget'] + files_list)
+        subprocess.call(['dmget'] + archive_files)
```
Comment (spencerkclark) [inline aospy/utils/io.py:165]: I was finding that the test suite would hang if `dmget` was called on an existing file that was not stored in `/archive`.  I was able to reproduce this outside of aospy, so it has nothing to do with the changes I made to allow for the recursive calculation of variables; they must have made some updates to `dmget`.  The logic here filters out files that aren't stored on archive before passing them to `dmget`.

File: aospy/data_loader.py [hunk 2]
```diff
@@ -296,7 +296,7 @@ def recursively_compute_variable(self, var, start_date=None, end_date=None,
             data = [self.recursively_compute_variable(
                 v, start_date, end_date, time_offset, **DataAttrs)
                     for v in var.variables]
-            return var.func(*data)
+            return var.func(*data).rename(var.name)
```
Comment (spencerkclark) [inline aospy/data_loader.py]: Indeed, good suggestion to add a test for this.  One minor change and things work smoothly!

**Review by spencerkclark — COMMENTED** [review]
Thanks @spencerahill!  I agree this should be highlighted in the documentation.  It will certainly simplify my object library quite a bit.  I'll add a full docstring to `recursively_compute_variable` and an example to the official docs sometime later today or tomorrow.

**Comment by spencerahill** [comment]
All sounds/looks good.  Just ping me when you're ready for final review.

**Comment by spencerkclark** [comment]
I guess I spoke too soon in https://github.com/spencerahill/aospy/issues/259; that's the failing test here :)

@spencerahill anyway, whenever you get a chance, this should be ready for a final review!

**Comment by spencerahill** [comment]
>I guess I spoke too soon in #259; that's the failing test here :)

Bummer.  Still, it seems like the frequency of the failures has gone down...maybe it's on its way toward going away completely.

In it goes!  Thanks @spencerkclark, this is great.

**Pull request merged**